exporter/elasticsearchexporter/integrationtest/exporter_test.go

```diff
@@ -0,0 +1,170 @@
+	"go.opentelemetry.io/collector/pdata/plog/plogotlp"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+func TestExporter(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		// restartCollector restarts the OTEL collector. Restarting
+		// the collector allows durability testing of the ES exporter
+		// based on the OTEL config used for testing.
+		restartCollector bool
+		mockESFailure    bool
+	}{
+		{name: "basic"},
+		{name: "es_intermittent_failure", mockESFailure: true},
+		/* Below tests should be enabled after #30792 is fixed
+		{name: "collector_restarts", restartCollector: true},
+		{name: "collector_restart_with_es_intermittent_failure", mockESFailure: true, restartCollector: true},
+		*/
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			runner(t, tc.restartCollector, tc.mockESFailure)
+		})
+	}
+}
+
+func runner(t *testing.T, restartCollector, mockESFailure bool) {
+	t.Helper()
+
+	cfg := loadConfig(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	flushIvl := time.Second
+	mockES := newMockESClient(t, cfg.Debug)
+	collector := newTestCollector(t, cfg, mockES.ServerURL, flushIvl)
+
+	var g errgroup.Group
+	cancelRun := runTestCollectorWithWait(ctx, t, collector, &g)
+
+	if mockESFailure {
+		mockES.SetReturnStatusCode(http.StatusServiceUnavailable)
+		if !restartCollector {
+			// If restartCollector is not set then recover ES service
+			// after a few failed bulk requests otherwise ensure that
+			// both collector and ES are unavailable for the same
+			// duration considering worst case scenario. ES recovers
+			// after a certain number of bulk requests have happened.
+			var bulkCalls atomic.Int32
+			mockES.SetOnBulkRequest(func(_ esutil.BulkIndexerResponse) {
+				if count := bulkCalls.Add(1); count == 10 {
+					mockES.SetReturnStatusCode(http.StatusOK)
+				}
+			})
+		}
+	}
+
+	count := sendLogs(t, cfg.GRPCEndpoint, "batch_1", 500, 10) // total=logs*agents
+
+	if restartCollector {
+		// Restart the collector after all data is sent to the collector.
+		// Note that the log count should be high to ensure that the collector
+		// stops before all data is shipped to Elasticsearch.
+		require.NoError(t, collector.Recreate())
+		cancelRun()
+		// Ensure that the previous collector instance is shut down and then
+		// run the recreated collector.
+		require.NoError(t, g.Wait())
+		runTestCollectorWithWait(ctx, t, collector, &g)
+		// Recover ES if it was unavailable.
+		mockES.SetReturnStatusCode(http.StatusOK)
+	}
+
+	count += sendLogs(t, cfg.GRPCEndpoint, "batch_2", 500, 10) // total=logs*agents
+
+	assert.Eventually(
+		t, func() bool {
+			resp, err := mockES.Count(mockES.Count.WithIndex(cfg.ESLogsIndex))
+			require.NoError(t, err)
+			defer resp.Body.Close()
+
+			body, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+
+			result := gjson.GetBytes(body, "count")
+			return result.Int() == int64(count)
```

Comment:
Depending on reliability guarantees, they may not be equal. e.g. for at-least-once, it should be `result.Int() >= int64(count)`. I believe this eventually needs to be changed to `result.Int() >= int64(count)` but I don't have a strong opinion on whether it needs to be in this PR.

Reply:
The current tests don't test for delivery guarantees and the tests are not designed to simulate at-least-once guarantees. Maybe in future, we will need this but it is not required at the moment.